### PR TITLE
Accept ApplicationSnapshot inputs

### PR DESCRIPTION
**Pull request by caugello**
Validates inputs from string or file agains its schema.

### CODE REVIEW

File: cmd/internal/applicationsnapshot/validation.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+type Report struct {
+	Success bool    `json:"success"`
+	Images  []Image `json:"images"`
+}
+
+const SCHEMA = "https://raw.githubusercontent.com/redhat-appstudio/release-service/main/config/crd/bases/appstudio.redhat.com_applicationsnapshots.yaml"
+
+// Returns an error on failure.
+func Validate(jsonstring string) error {
```
Comment (robnester-rh) [inline cmd/internal/applicationsnapshot/validation.go:47]: Functions require docstrings in the form of `// Validate ...`

File: cmd/eval.go [hunk 2]
```diff
@@ -22,16 +22,6 @@ import (
...
-var arguments = struct {
-	policyConfiguration string
-	imageRef            string
-	publicKey           string
-	rekorURL            string
-	strict              bool
-}{
-	rekorURL: "https://rekor.sigstore.dev/",
-}
-
```
Comment (zregvart) [inline cmd/eval.go:34]: I don't quite understand how is this compiling, the code below uses `arguments`

File: cmd/verify.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		report, err := applicationsnapshot.GenerateReport(snapshot)
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+
+		fmt.Println(report)
+		os.Exit(0)
+	},
```
Comment (zregvart) [inline cmd/verify.go:83]: In eval we do `RunE` instead and return the err, this is not nice and we have [HACBS-633](https://issues.redhat.com/browse/HACBS-633) to clean that up...

**Comment by zregvart** [comment]
Just to note [go generate build](https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source) tag that can be used to fetch the CRD YAML

File: cmd/eval.go (hunk 2 above)
Comment (caugello) [reply cmd/eval.go:34]: I think it's because it's the same package.

File: cmd/verify.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
+/*
```
[the remark this answers is not in the document]
Comment (caugello) [reply cmd/verify.go:1]: I am not sure I understand, it's under cmd.

File: cmd/verify.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// verifyCmd represents the verify command
+var verifyCmd = &cobra.Command{
+	Use:   "verify",
+	Short: "Change me",
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var jsonstr string
+		// Fail if input and filepath are used.
+		if arguments.input != "" && arguments.filepath != "" {
+			fmt.Println("Use either --input or --filepath")
+			os.Exit(1)
```
[the remark this answers is not in the document]
Comment (caugello) [reply cmd/verify.go]: Yes, updated it.

File: cmd/eval.go (hunk 2 above)
Comment (zregvart) [reply cmd/eval.go:34]: Right I missed that these got moved to `root.go`, can we keep them here instead, the intent is that these arguments are for the `eval` functionality, trying to share this struct between two sub-comands will lead to unexpected behavior (e.g. nil reference) on trying to access a field that was not bound.

File: cmd/internal/applicationsnapshot/applicationsnapshot_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+//go:embed test_crd.yaml
+var crdYAML string
+
+func fetchStub() (*os.File, error) {
+	file, err := ioutil.TempFile(os.TempDir(), "ec-cli-stubcrd")
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/applicationsnapshot_test.go:32]: Not really important, this is an equivalent:

```suggestion
	file, err := ioutil.TempFile("", "ec-cli-stubcrd")
```

File: cmd/internal/applicationsnapshot/applicationsnapshot_test.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	if err != nil {
+		return nil, err
+	}
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/applicationsnapshot_test.go:35]: File should be closed

```suggestion
	}
	defer file.Close()
```

File: cmd/internal/applicationsnapshot/source.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	"context"
+	"io/ioutil"
+	"os"
+
+	"github.com/hashicorp/go-getter"
+)
+
+const SCHEMA = "https://raw.githubusercontent.com/redhat-appstudio/release-service/main/config/crd/bases/appstudio.redhat.com_applicationsnapshots.yaml"
+
+type fetcherFn func() (*os.File, error)
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source.go:29]: Perhaps we should return Reader here instead?
```suggestion
type fetcherFn func() (*io.Reader, error)
```

File: cmd/internal/applicationsnapshot/report.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Report the states of images from the snapshot
+func Report(snapshot *rs.ApplicationSnapshotSpec) (string, error) {
+	output := report{Success: true}
+
+	for _, image := range snapshot.Images {
+		item := Image{
+			Violations: []string{},
+			Success:    true,
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/report.go:43]: Shouldn't we compute values for these?

File: cmd/internal/applicationsnapshot/report_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"strings"
+	"testing"
+
+	rs "github.com/redhat-appstudio/release-service/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+)
+
+//go:embed test_snapshot.json
+var jsonstr string
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/report_test.go:31]: Perhaps a more indicative name?

```suggestion
var testSnapshot string
```

File: cmd/internal/applicationsnapshot/report_test.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+          "violations": [],
+          "success": true
+        }
+      ]
+    }
+  `
+
+	expected := strings.ReplaceAll(strings.ReplaceAll(string(output), "\n", ""), " ", "")
+	report, _ := Report(snapshot)
+	assert.Equal(t, expected, report)
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/report_test.go:68]: Could use [`assert.JSONEq`](https://pkg.go.dev/github.com/stretchr/testify/assert#JSONEq) or [jsonassert](https://github.com/kinbiko/jsonassert) here

File: cmd/internal/applicationsnapshot/source.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	return schemaSource{
+		fetch: fetch,
+	}
+}
+
+// fetch the latest CRD and return the file
+//go:generate build
+func fetch() (*os.File, error) {
+	// Create a tempfile for the CRD
+	file, err := ioutil.TempFile(os.TempDir(), "ec-cli-applicationsnapshot")
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source.go:46]: For consistency of nitpickery

```suggestion
	file, err := ioutil.TempFile("", "ec-cli-applicationsnapshot")
```

File: cmd/internal/applicationsnapshot/source.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Download the CRD
+	client := getter.Client{
+		Ctx:      context.Background(),
+		Src:      SCHEMA,
+		Dst:      file.Name(),
+		Insecure: false,
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source.go:56]: Should be the default

```suggestion
```

File: cmd/internal/applicationsnapshot/source.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+		Getters: map[string]getter.Getter{
+			"https": &getter.HttpGetter{},
+		},
+	}
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source.go:60]: Simpler version would be:

```go
	getter.Get(file.Name(), SCHEMA, getter.WithContext(ctx))
```

`ctx` can be obtained from the Command instance (`cmd.Context()`)

File: cmd/internal/applicationsnapshot/source.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+
+	"github.com/hashicorp/go-getter"
+)
+
+const SCHEMA = "https://raw.githubusercontent.com/redhat-appstudio/release-service/main/config/crd/bases/appstudio.redhat.com_applicationsnapshots.yaml"
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source.go:27]: I think the getter client can be forced here, so `https` works out of the box below to make the invocation a bit simpler.

```suggestion
const SCHEMA = "http::https://raw.githubusercontent.com/redhat-appstudio/release-service/main/config/crd/bases/appstudio.redhat.com_applicationsnapshots.yaml"
```

File: cmd/internal/applicationsnapshot/source_test.go [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_NewSchemaSource(t *testing.T) {
+	source := NewSchemaSource()
+	assert.IsType(t, schemaSource{}, source)
+}
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/source_test.go:28]: We want to test behavior that is not enforced by the compiler, but this helps with test coverage...

File: cmd/internal/applicationsnapshot/applicationsnapshot_test.go [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+	_ "embed"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+//go:embed test_crd.yaml
+var crdYAML string
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/applicationsnapshot_test.go:29]: Perhaps name this after the file?

```suggestion
var testCRD string
```

File: cmd/internal/applicationsnapshot/validation.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+	// Struct to unmarshal the OpenAPIV3Schema required for validation.
+	var schemaProps struct {
+		Properties struct {
+			Spec *crd.JSONSchemaProps `json:"spec"`
+		} `json:"properties"`
+	}
+
+	//TODO: there must be a better way to avoid those 3 steps
+	// Marshal the openapi schema to JSON
+	openapiV3Schema, err := json.Marshal(s.Spec.Versions[0].Schema.OpenAPIV3Schema)
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/validation.go:40]: We need to be careful about which version we use here, version at index 0 could be the one that is not served any more, that is I think we should filter on `CustomResourceDefinitionVersion.Served == true`

File: cmd/internal/applicationsnapshot/validation_test.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+			name: "Missing attribute",
+			input: `{
+          "images": [
+            {
+              "component": "spam",
+              "pullSpec": "quay.io/caf/spam@sha256:123…"
+            }
+          ]
+        }
+      `,
```
Comment (zregvart) [inline cmd/internal/applicationsnapshot/validation_test.go:59]: Nitpick, spaces vs tabs...

File: cmd/root.go [hunk 20]
```diff
@@ -33,6 +33,18 @@ var rootCmd = &cobra.Command{
...
+	policyConfiguration string
+	imageRef            string
+	publicKey           string
+	rekorURL            string
+	strict              bool
+	input               string
+	filepath            string
+}{
+	rekorURL: "https://rekor.sigstore.dev/",
+}
```
Comment (zregvart) [inline cmd/root.go:46]: Let's have a version of this struct for `eval` and a version for `verify`

**Review by zregvart — COMMENTED** [review]
It seems to me that we go to much trouble to validate the schema of the `ApplicationSnapshot`, we could have the same benefit of schema validation and more strict validation of individual fields by unmarshalling to an expected `struct` and validating the fields ourselves.

File: cmd/eval.go (hunk 2 above)
Comment (caugello) [reply cmd/eval.go:34]: We can, issue is that eval and verify are under the `cmd` package, any `var` will be accessible for both

Comment (caugello) [reply cmd/eval.go:34]: If that the case, we prefix them within the struct to have `arguments.eval.imageRef` wdyt ?

File: cmd/internal/applicationsnapshot/report.go (hunk 9 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/report.go:43]: We should, but in a later story. Scope is input validation and display the report from input and with above values.

File: cmd/internal/applicationsnapshot/source.go (hunk 15 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/source.go:27]: Weirdly, without the config I get ` no source URL was returned`

File: cmd/internal/applicationsnapshot/source.go (hunk 14 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/source.go:60]: This doesn't seem to work. getting ` no source URL was returned`

File: cmd/internal/applicationsnapshot/validation.go (hunk 18 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/validation.go]: Correct, changed it.

File: cmd/internal/applicationsnapshot/source.go (hunk 12 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/source.go:47]: With `io.reader` we don't need a tmpfile anymore.

File: cmd/internal/applicationsnapshot/source_test.go (hunk 16 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/source_test.go:28]: Yeah, that's a coverage test.

File: cmd/internal/applicationsnapshot/source.go (hunk 8 above)
Comment (caugello) [reply cmd/internal/applicationsnapshot/source.go]: Yup, changed it.

File: cmd/root.go (hunk 20 above)
Comment (caugello) [reply cmd/root.go]: Agreed, structs accessible trough `arguments`

File: cmd/eval.go (hunk 2 above)
Comment (zregvart) [reply cmd/eval.go]: I think we should do something that @robnester-rh suggested and implemented in the validate pipeline PR: https://github.com/hacbs-contract/ec-cli/pull/15/files#diff-6701a60d95525f3d714ff656fdc86ef5dc9753df9e79b60f0a9a7d2be4d884c6R34-R48

But we can refactor this after this PR to have this done first...

Comment (caugello) [reply cmd/eval.go]: Yup, made the change

Comment (caugello) [reply cmd/eval.go]: Moved it to eval, and got rid of verify

File: cmd/eval.go [hunk 21]
```diff
@@ -17,80 +17,143 @@ limitations under the License.
...
+
+					jsonstr = string(file)
+				}
+
+				// Input handler
+				if arguments.input != "" {
+					jsonstr = arguments.input
+				}
+				// Unmarshall json into struct, exit on failure
+				if err := json.Unmarshal([]byte(jsonstr), &snapshotSpec); err != nil {
```
Comment (zregvart) [inline cmd/eval.go:74]: Hmm, is this `**appstudioshared.ApplicationSnapshotSpec`? Looks odd given _If v is nil or not a pointer, Unmarshal returns an InvalidUnmarshalError._ from the docs of [json.Unmarshal](https://pkg.go.dev/encoding/json#Unmarshal).
I think we're missing an unit test for this to make sure.

File: cmd/eval.go [hunk 22]
```diff
@@ -17,80 +17,143 @@ limitations under the License.
...
+				if snapshotSpec == nil {
+					return errors.New("ApplicationSnapshot input could not be unmarshalled.")
+				}
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			//TODO: refactor once we handle application snapshots
+			if arguments.input != "" || arguments.filepath != "" {
```
Comment (zregvart) [inline cmd/eval.go:87]: Perhaps we could check `snapshotSpec != nil` here?

**Pull request merged**